api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -243,43 +243,58 @@ private DbUser initializeUserIfNeeded() {
-    // created Google accounts for. This can be used subsequently to delete orphaned accounts.
-
-    // We store this information in our own database so that:
-    // 1) we can support bring-your-own account in future (when we won't be using directory service)
-    // 2) we can easily generate lists of researchers for the storefront, without joining to Google
-
-    // It's possible for the profile information to become out of sync with the user's Google
-    // profile, since it can be edited in our UI as well as the Google UI,  and we're fine with
-    // that; the expectation is their profile in AofU will be managed in AofU, not in Google.
-
-    DbUser user =
-        userService.createUser(
-            profile.getGivenName(),
-            profile.getFamilyName(),
-            googleUser.getPrimaryEmail(),
-            profile.getContactEmail(),
-            profile.getCurrentPosition(),
-            profile.getOrganization(),
-            profile.getAreaOfResearch(),
-            profile.getProfessionalUrl(),
-            profile.getDegrees(),
-            FROM_CLIENT_ADDRESS.apply(profile.getAddress()),
-            demographicSurveyMapper.demographicSurveyToDbDemographicSurvey(
-                profile.getDemographicSurvey()),
-            verifiedInstitutionalAffiliationMapper.modelToDbWithoutUser(
-                profile.getVerifiedInstitutionalAffiliation(), institutionService));
+    DbUser user;
+    try {
+      user =
+          userService.createUser(
+              profile.getGivenName(),
+              profile.getFamilyName(),
+              googleUser.getPrimaryEmail(),
+              profile.getContactEmail(),
+              profile.getCurrentPosition(),
+              profile.getOrganization(),
+              profile.getAreaOfResearch(),
+              profile.getProfessionalUrl(),
+              profile.getDegrees(),
+              FROM_CLIENT_ADDRESS.apply(profile.getAddress()),
+              demographicSurveyMapper.demographicSurveyToDbDemographicSurvey(
+                  profile.getDemographicSurvey()),
+              verifiedInstitutionalAffiliationMapper.modelToDbWithoutUser(
+                  profile.getVerifiedInstitutionalAffiliation(), institutionService));
+    } catch (Exception e) {
+      // If the creation of a User row in the RW database fails, we want to attempt to remove the
+      // G Suite account to avoid having an orphaned account with no record in our database.
+      log.severe(
+          String.format(
+              "An error occurred when creating DbUser for %s. Attempting to delete "
+                  + "orphaned G Suite account",
+              gSuiteUsername));
+      try {
+        directoryService.deleteUser(gSuiteUsername);
+        log.severe("Orphaned G Suite account has been deleted.");
+      } catch (Exception e2) {
+        log.severe(
+            String.format(
+                "Orphaned G Suite account %s could not be deleted. "
```

Comment:
Should include at least the `e2` message here, for additional context

Reply:
Good idea – done.

Reply:
This is the first time I've seen a nested try/catch. I'd probably factor out a `deleteOrphanedUser()` method.

Also, is there a general pattern for chaining methods like this into a transaction? One thing you could do is have an initialization_status column that's set to one state after creating the row and another when creating the GSuite user. Then just have a cron harvest the ones that are over 24 hours old with no gsuite account.

Reply:
Factoring out wouldn't hurt, but for one line of code I don't think it helps too much either. That would also make it more awkward to return the original exception (which I believe is the behavior we want, so that the client is guaranteed to receive the originating error message rather than a misleading error message about a potentially failed G Suite cleanup attempt).

Frankly I'm okay with this section being a little bit awkward, since we're entering into the realm of extra-paranoid programming to catch uncommon-but-important errors that we've seen in production. My hope is that liberal logging provides some production visibility if & when this happens again, and liberal commenting will provide hints to future devs.

More generally, TDR and the rest of Terra have aligned on Dan Dietterich's home-built [Stairway](https://github.com/databiosphere/stairway) library which implements the [Saga](https://blog.couchbase.com/saga-pattern-implement-business-transactions-using-microservices-part/#:~:text=The%20SAGA%20Pattern&text=A%20saga%20is%20a%20sequence,completion%20of%20the%20previous%20one.) pattern for distributed transactions. It definitely adds some overhead to implementing simple flows, but it provides the necessary state management and hooks to reliably roll back at any point in a long-running operation. Another equivalent library out there is [Cadence](https://github.com/uber/cadence), open-sourced by Uber.